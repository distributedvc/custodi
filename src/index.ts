import { utm } from '@distributed/utm';
import Cookie from 'cookie-universal';

interface CustodiParams {
  query: string;
  timeout?: number;
}

export function custodi({ timeout = 30, query }: CustodiParams): void {
  const utms = utm(query);
  const now = new Date();
  const cookies = Cookie();

  function updateCookie(key: string, value: string, expires: Date): void {
    if (value) {
      cookies.set(key, value, {
        path: '/',
        expires,
      });
    }
  }

  function updateExpiration(key: string, expires: Date): void {
    const existingValue = cookies.get(key);

    if (existingValue) {
      cookies.set(key, existingValue, {
        expires,
      });
    }
  }

  function removeCookies(): void {
    for (const key in utms) {
      cookies.remove(key);
    }
  }

  const expires = new Date(now.setMinutes(now.getMinutes() + timeout));

  let removed = false;

  for (const key in utms) {
    const param = key;

    updateExpiration(param, expires);

    const value = utms[key];

    if (value) {
      if (!removed) {
        removeCookies();
        removed = true;
      }
      updateCookie(param, value as string, expires);
    }
  }
}
