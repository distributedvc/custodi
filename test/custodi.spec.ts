import { custodi } from '../src';
import Cookie from 'cookie-universal';

describe('Custodi', (): void => {
  it('should get utms cookies', async (): Promise<void> => {
    custodi({
      query:
        '?utm_source=google&utm_campaign=name&utm_medium=medium&utm_content=content&utm_term=term&foo=bar&baz=foo',
    });

    const cookies = Cookie();
    const utmCookies = cookies.getAll();

    expect(utmCookies).toMatchSnapshot();
  });
});
