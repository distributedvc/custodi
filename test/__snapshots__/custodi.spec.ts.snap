// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custodi should extract correct utms from the query 1`] = `
Object {
  "utm_campaign": "name",
  "utm_content": "content",
  "utm_medium": "medium",
  "utm_source": "google",
  "utm_term": "term",
}
`;

exports[`Custodi should get utms cookies 1`] = `
Object {
  "utm_campaign": "name",
  "utm_content": "content",
  "utm_medium": "medium",
  "utm_source": "google",
  "utm_term": "term",
}
`;
